File: engine/dynamic/render_opengl/src/renderer/2d/scene_compiler.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/collections.hpp"

#include "argus/render/defines.hpp"
#include "argus/render/util/object_processor.hpp"
#include "argus/render/2d/scene_2d.hpp"

#include "internal/render_opengl/renderer/shader_mgmt.hpp"
#include "internal/render_opengl/renderer/2d/object_proc_impl.hpp"
#include "internal/render_opengl/renderer/2d/scene_compiler.hpp"
#include "internal/render_opengl/state/processed_render_object.hpp"
#include "internal/render_opengl/state/render_bucket.hpp"
#include "internal/render_opengl/state/scene_state.hpp"

namespace argus {
    static BucketKey _get_bucket_key(ProcessedRenderObject &processed_obj) {
        return BucketKey { processed_obj.material_res.uid, processed_obj.atlas_stride, processed_obj.z_index };
    }

    static void _create_obj_ubo(RenderBucket &bucket) {
        bucket.obj_ubo = BufferInfo::create(GL_UNIFORM_BUFFER, SHADER_UBO_OBJ_LEN, GL_STATIC_DRAW, false);

        float stride[2] = { bucket.atlas_stride.x, bucket.atlas_stride.y };
        bucket.obj_ubo.write(stride, sizeof(stride), SHADER_UNIFORM_OBJ_UV_STRIDE_OFF);
    }

    static void _handle_new_obj(Scene2DState &scene_state, ProcessedRenderObject &processed_obj) {
        RenderBucket *bucket;
        auto key = _get_bucket_key(processed_obj);
        auto existing_bucket_it = scene_state.render_buckets.find(key);
        if (existing_bucket_it != scene_state.render_buckets.end()) {
            bucket = existing_bucket_it->second;
        } else {
            bucket = &RenderBucket::create(processed_obj.material_res, processed_obj.atlas_stride,
                    processed_obj.z_index);
            scene_state.render_buckets[key] = bucket;
            _create_obj_ubo(*bucket);
        }

        bucket->objects.push_back(&processed_obj);
        bucket->needs_rebuild = true;

        processed_obj.newly_created = false;
    }

    static void _handle_stale_obj(Scene2DState &scene_state, ProcessedRenderObject &processed_obj) {
        deinit_object_2d(processed_obj);

        // we need to remove it from its containing bucket and flag the bucket for a rebuild
        auto *bucket = scene_state.render_buckets[_get_bucket_key(processed_obj)];
        remove_from_vector(bucket->objects, &processed_obj);
        bucket->needs_rebuild = true;

        processed_obj.~ProcessedRenderObject();
    }

    void compile_scene_2d(const Scene2D &scene, Scene2DState &scene_state) {
        process_objects_2d(scene, scene_state.processed_objs, create_processed_object_2d, update_processed_object_2d,
                &scene_state);

        for (auto it = scene_state.processed_objs.begin(); it != scene_state.processed_objs.end();) {
            auto &processed_obj = *reinterpret_cast<ProcessedRenderObject *>(it->second);
            if (processed_obj.newly_created) {
                _handle_new_obj(scene_state, processed_obj);
            } else if (!processed_obj.visited) {
                // wasn't visited this iteration, must not be present in the scene graph anymore
                _handle_stale_obj(scene_state, processed_obj);
                it = scene_state.processed_objs.erase(it);
                continue;
            }

            processed_obj.visited = false;
            ++it;
        }
    }
}

File: engine/modules/render_opengl/src/twod/scene_compiler.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::collections::BTreeMap;
use argus_render::constants::*;
use argus_render::util::process_objects_2d;
use crate::aglet::*;
use crate::state::*;
use crate::twod::{deinit_object_2d, process_object};
use crate::util::buffer::GlBuffer;

fn get_bucket_key(processed_obj: &ProcessedObject) -> RenderBucketKey {
    RenderBucketKey::new(
        processed_obj.material_res.get_prototype().uid.clone(),
        &processed_obj.atlas_stride,
        processed_obj.z_index,
        processed_obj.light_opacity
    )
}

fn create_obj_ubo(bucket: &mut RenderBucket) {
    bucket.obj_ubo = Some(
        GlBuffer::new(GL_UNIFORM_BUFFER, SHADER_UBO_OBJ_LEN as usize, GL_STATIC_DRAW, true, false)
    );
    let ubo = &bucket.obj_ubo.as_ref().unwrap();

    // we assume that these values will never change

    let stride = [bucket.atlas_stride.x, bucket.atlas_stride.y];
    ubo.write_val(&stride, SHADER_UNIFORM_OBJ_UV_STRIDE_OFF as usize);

    let light_opacity = bucket.light_opacity;
    ubo.write_val(&light_opacity, SHADER_UNIFORM_OBJ_LIGHT_OPACITY_OFF as usize);
}

fn handle_new_obj(
    render_buckets: &mut BTreeMap<RenderBucketKey, RenderBucket>,
    processed_obj: &mut ProcessedObject
) {
    let key = get_bucket_key(processed_obj);
    let bucket = render_buckets.entry(key)
        .or_insert_with(|| {
            let mut bucket = RenderBucket::create(
                processed_obj.material_res.clone(),
                processed_obj.atlas_stride,
                processed_obj.z_index,
                processed_obj.light_opacity
            );
            create_obj_ubo(&mut bucket);
            bucket
        });

    processed_obj.newly_created = false;

    bucket.objects.push(processed_obj.obj_handle);
    bucket.needs_rebuild = true;
}

fn handle_stale_obj(
    render_buckets: &mut BTreeMap<RenderBucketKey, RenderBucket>,
    processed_obj: &mut ProcessedObject
) {
    deinit_object_2d(processed_obj);

    // we need to remove it from its containing bucket and flag the bucket for a rebuild
    let bucket = render_buckets.get_mut(&get_bucket_key(processed_obj)).unwrap();
    bucket.objects
        .remove(bucket.objects.iter().position(|h| h == &processed_obj.obj_handle).unwrap());
    bucket.needs_rebuild = true;
}

pub(crate) fn compile_scene_2d(renderer_state: &mut RendererState, scene_id: impl AsRef<str>) {
    process_objects_2d(
        scene_id.as_ref(),
        process_object,
        renderer_state,
    );

    let scene_state = renderer_state.scene_states_2d
        .entry(scene_id.as_ref().to_string())
        .or_insert_with(|| {
            Scene2dState {
                scene_id: scene_id.as_ref().to_string(),
                ubo: None,
                render_buckets: Default::default(),
                processed_objs: Default::default(),
            }
        });

    scene_state.processed_objs.retain(|_, obj| {
        if obj.newly_created {
            handle_new_obj(&mut scene_state.render_buckets, obj);
        } else if !obj.visited {
            // wasn't visited this iteration, must not  be present in the scene graph anymore
            handle_stale_obj(&mut scene_state.render_buckets, obj);
            return false;
        }

        obj.visited = false;
        return true;
    });
}
